### PR TITLE
sql: remove castInfo and add tests for cast contexts

### LINKED ISSUE #74577: sql: disallow casts from INT->VARBIT
Casts from `INT` to `VARBIT` should not be allowed because they are not supported in Postgres and they have unexpected behavior in CRDB.

In Postgres:

```
marcus=# SELECT 1::VARBIT;
ERROR:  42846: cannot cast type integer to bit varying
LINE 1: SELECT 1::VARBIT;
                ^
LOCATION:  transformTypeCast, parse_expr.c:2652

marcus=# SELECT 1::INT8::VARBIT;
ERROR:  42846: cannot cast type bigint to bit varying
LINE 1: SELECT 1::INT8::VARBIT;
                      ^
LOCATION:  transformTypeCast, parse_expr.c:2652
```

In CRDB v21.2.3:

```
defaultdb> SELECT 1::VARBIT;
  varbit
----------

(1 row)

defaultdb> SELECT 1::INT8::VARBIT;
  varbit
----------

(1 row)
```

### PR DESCRIPTION
**Pull request by mgartner**
#### sql: rename pg_cast_provolatile_dump.csv to pg_cast_dump.csv

Release note: None

#### sql: remove castInfo

`castMap` now contains all volatility information and `castInfo` is no
longer needed. `castInfo` has been removed.

For backwards compatibility, some invalid casts between VARBIT and
integer types have been added to `castMap`. These casts have unexpected
behavior before and after this commit. They are not supported by
Postgres.  They should be disallowed in the future.
See #74577 and 74580

Casts between tuple and string types are now dynamically handled in
`lookupCast` support casts with named record types that have different
OIDs than `oid.T_record`.

Casts from OID and REG* types to INT2 are now allowed to maintain
backward compatibility.

Release note: None

#### sql: update pg_cast_dump.csv

This commit updates the `pg_cast_dump.csv` file with new rows for casts
from the JSONB type (OID 3802). It also adds the Postgres version as a
column to annotate the version of Postgres that the CSV was generated
from. This is important because the `pg_cast` table can change from
version to version. This CSV was generated from Postgres version 13.5.

Release note: None

#### sql: consistently order pg_cast_dump.csv

Release note: None

#### sql: check cast contexts against Postgres's pg_cast table

This commit includes the `pg_cast.castcontext` column in
`pg_cast_dump.csv` and uses it to validate the `maxContext` field in
`castMap`. Note that Postgres's `pg_cast` table does not include all
possible casts, such as automatic I/O conversions, so this test is not
comprehensive.

Release note: None

### CODE REVIEW

**Comment by mgartner** [comment]
The commit that changes some of the casts' volatilities to leak-proof is somewhat risky, as I mentioned in the commit message:
```
I'd like to point out that, while it is important that our cast
volatilities match Postgres for the sake on consistency, it's more
important that our volatilities match the behavior of our cast
operators. If we mark a cast as leak-proof, it means that that cast will
never throw an error. If the cast actually can throw an error, then
there will likely be correctness bugs as a result.
```
I'm also not convinced there's a lot of benefit. It technically makes use more consistent with Postgres, but from a user's perspective there's not a huge difference between immutable and leak-proof, and it might allow some minor optimizations in some cases. I could revert those changes if we don't have high enough confidence that our version of those casts are indeed leak-proof.

Maybe @RaduBerinde intentionally diverged from Postgres when he added the leak-proof column in #49680? I can't find any explanation of why the proleakproof values he added to the pg_cast dump CSV do not match the actual values in pg_cast, so I can't be sure.

**Review by RaduBerinde — COMMENTED** [review]
:lgtm: 

[nit] The `sql: remove castInfo` commit description omits the functional change that fixes that issue.

> Maybe @RaduBerinde intentionally diverged from Postgres when he added the leak-proof column in #49680? I can't find any explanation of why the proleakproof values he added to the pg_cast dump CSV do not match the actual values in pg_cast, so I can't be sure.

I only see `false` values (tried on 10.2 and 12.9):
```
psql (12.9 (Ubuntu 12.9-0ubuntu0.20.04.1))
radu=# select pg_cast.* from pg_catalog.pg_cast, pg_catalog.pg_proc where pg_proc.oid = castfunc and proleakproof;
 oid | castsource | casttarget | castfunc | castcontext | castmethod 
-----+------------+------------+----------+-------------+------------
(0 rows)
```

What version did you use? I wish we could put the version somewhere, but there are no comments in CSV files.. At least make sure you include the version in the description for all commits that modify the csv dump.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74578#-:-Msqndbr5NJvecwxJWeI:bajcqsc)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan and @rytaft)

File: pkg/sql/sem/tree/cast_test.go [hunk 1]
```diff
@@ -52,13 +52,19 @@ func TestCastsVolatilityMatchesPostgres(t *testing.T) {
...
+//     SELECT
+//       c.castsource,
+//       c.casttarget,
+//       p.provolatile,
+//       p.proleakproof,
+//       c.castcontext
 //     FROM pg_cast c JOIN pg_proc p ON (c.castfunc = p.oid)
 //     WHERE
 //       c.castsource NOT IN (SELECT t FROM ignored_types)
 //       AND c.casttarget NOT IN (SELECT t FROM ignored_types)
```
Comment (otan) [inline pkg/sql/sem/tree/cast_test.go:64]: wanna do an order by here?

File: pkg/sql/logictest/testdata/logic_test/cast [hunk 2]
```diff
@@ -777,3 +777,26 @@ query T
 EXECUTE s73450_c('foo')
 ----
 f
+
+# Regression test for #74577. Casts between integer types and VARBIT should not
```
Comment (otan) [inline pkg/sql/logictest/testdata/logic_test/cast:781]: this probably needs a deprecation process as well, unfortunately.
(detect any usages, and block migration...)

File: pkg/sql/sem/tree/cast_test.go [hunk 3]
```diff
@@ -221,16 +117,68 @@ func TestCastsFromUnknown(t *testing.T) {
...
+		types.TimestampTZArray,
+		types.TimeTZ,
+		types.TimeTZArray,
+		types.Uuid,
+		types.UUIDArray,
+		types.VarBit,
+		types.VarBitArray,
+		types.VarChar,
+		types.MakeArray(types.VarChar),
+		types.Void,
```
Comment (otan) [inline pkg/sql/sem/tree/cast_test.go:221]: hmm, how do we future proof this against adding new types?

File: pkg/sql/logictest/testdata/logic_test/cast (hunk 2 above)
Comment (mgartner) [reply pkg/sql/logictest/testdata/logic_test/cast:781]: I'd like to just fix it before the release. My attempt: https://github.com/cockroachdb/cockroach/pull/74583

**Review by mgartner — COMMENTED** [review]
> commit description omits the functional change that fixes that issue.
 
There was no functional change, just a correct list of valid casts. I've decided to add the invalid casts to the `castMap` for now. The bug remains unresolved.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74578#-:-Mt4NtDw2Gu0sfIaPB-E:biqp6b7)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @otan and @rytaft)

___
*[pkg/sql/sem/tree/cast_test.go, line 175 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74578#-Msq5Rt2CHlopfKo8cGs-r2-175:-Mt4OJpT8Nf5mZuy3CbR:ba3bc50) ([raw file](https://github.com/cockroachdb/cockroach/blob/9116bbd988e6a99125ea01d5ff125600f3a9a544/pkg/sql/sem/tree/cast_test.go#L175)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

hmm, how do we future proof this against adding new types?
</blockquote></details>

New types will need to be added to this list. There's currently no type registry or global list we can rely on.
___
*[pkg/sql/sem/tree/cast_test.go, line 64 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/74578#-Msq5Rt2CHlopfKo8cGs-r4-64:-Mt4O9xD7FCkdvv9ZK5l:b-dkd9b1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5cc0c63a8135cc994d991120703a337f8ba84da/pkg/sql/sem/tree/cast_test.go#L64)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

wanna do an order by here?
</blockquote></details>

Added in a separate commit to make the diffs easier to review.
___
*[pkg/sql/logictest/testdata/logic_test/cast, line 781 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74578#-Msq5Rt4hEMho6EKuLF7-r2-781:-Mt4OS_y-pw2zXoPNM_Y:b-63kvmi) ([raw file](https://github.com/cockroachdb/cockroach/blob/9116bbd988e6a99125ea01d5ff125600f3a9a544/pkg/sql/logictest/testdata/logic_test/cast#L781)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I'd like to just fix it before the release. My first attempt: https://github.com/cockroachdb/cockroach/pull/74583
</blockquote></details>

Nevermind, I've decided to leave the broken casts in for now.

**Comment by mgartner** [comment]
> What version did you use? I wish we could put the version somewhere, but there are no comments in CSV files.. At least make sure you include the version in the description for all commits that modify the csv dump.

I used version 14.0. It looks like proleakproof change in this release:

> Mark built-in type coercion functions as leakproof where possible (Tom Lane)
> 
> This allows more use of functions that require type conversion in security-sensitive situations.

_https://www.postgresql.org/docs/release/14.0/_

I've regenerated the CSV from 13.5 to de-risk this change. I've also added the PG version as a column to the CSV - it was the only way I could figure to annotate the CSV with the version that was used to generate it.

**Review by RaduBerinde — COMMENTED** [review]
:lgtm: 

> I've also added the PG version as a column to the CSV - it was the only way I could figure to annotate the CSV with the version that was used to generate it.

Smart!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74578#-:-Mt4XNG-1rIkCu7be4Kt:bjyxzf0)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan and @rytaft)

**Comment by otan** [comment]
___
*[pkg/sql/sem/tree/cast_test.go, line 175 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74578#-Msq5Rt2CHlopfKo8cGs-r2-175:-Mt4aWwC1yY_mvHPRWxm:bgd0vuo) ([raw file](https://github.com/cockroachdb/cockroach/blob/9116bbd988e6a99125ea01d5ff125600f3a9a544/pkg/sql/sem/tree/cast_test.go#L175)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

New types will need to be added to this list. There's currently no type registry or global list we can rely on.
</blockquote></details>

can we use `types.OidToType`

**Comment by otan** [comment]
___
*[pkg/sql/sem/tree/cast_test.go, line 175 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74578#-Msq5Rt2CHlopfKo8cGs-r2-175:-Mt4aupwDqdnw-dDUtvj:btlh6j9) ([raw file](https://github.com/cockroachdb/cockroach/blob/9116bbd988e6a99125ea01d5ff125600f3a9a544/pkg/sql/sem/tree/cast_test.go#L175)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

can we use `types.OidToType`
</blockquote></details>

Or move this `SeedType` logic elsewhere, which includes arrays: https://github.com/cockroachdb/cockroach/blob/16f643d49a39d400e27fab196cd5072a534c5506/pkg/sql/randgen/type.go#L39-L58

can be a separate PR if you'd like...

**Review by otan — APPROVED** [review]
lgtm! assuming that types change will probably be a separate PR, but i guarantee it will be forgotten when new types are added as is.

**Review by mgartner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74578)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @otan and @rytaft)

___
*[pkg/sql/sem/tree/cast_test.go, line 175 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74578#-Msq5Rt2CHlopfKo8cGs-r2-175:-Mt4eEo43Sa62oJrQrn3:b3slqq9) ([raw file](https://github.com/cockroachdb/cockroach/blob/9116bbd988e6a99125ea01d5ff125600f3a9a544/pkg/sql/sem/tree/cast_test.go#L175)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

Or move this `SeedType` logic elsewhere, which includes arrays: https://github.com/cockroachdb/cockroach/blob/16f643d49a39d400e27fab196cd5072a534c5506/pkg/sql/randgen/type.go#L39-L58

can be a separate PR if you'd like...
</blockquote></details>

Doh. I realized this test was a remnant from an outed branch where these casts were added to `castMap`. They no longer exist in `castMap`, so this test isn't needed. See: https://github.com/cockroachdb/cockroach/blob/8bb30c9f7af5c8003c90db7337ef82bac19ff1e6/pkg/sql/sem/tree/cast.go#L1274-L1281

I've removed the test entirely.

**Comment by mgartner** [comment]
TFTRs! Going to merge this.

bors r+

**Pull request merged**